Dens query draft

jaredponn opened the pull request:
What was done:
- Database schema for DeNS UTxOs
- `dens-query` for syncing with ogmios + dumping relavent UTxOs to the aforementioned database
- tests to ensure that the database SQL queries are syntactically valid and "do what we expect them to do".
- Simple beginnings of an interface so clients can actually query the database
- A draft of a user manual. Read it [here](https://github.com/mlabs-haskell/DeNS/blob/jared/dens-query-draft/dens-query/doc/MANUAL.adoc)


TODO for future issues:
- [x] Allow it to sync from the last known point (it always syncs from the beginning of time)
- [ ] Check health endpoint / book keeping to see "how far synced" we are
- [ ] Compile cardano-serialization-lib with the `--weak-refs` flag s.t. it doesn't leak memory everywhere. For now, we manually free the memory allocated
- [x] Full tests with ogmios! See #29 
- [x] What is the story with the `class` vs the domain name? For now, it ignores the class completely. A: It looks at the transaction for the class

Commit:
Improved vm documentation + fixed postgres s.t. we can access it from
the host

Commit:
Improved documentation

Commit:
Fixed documentation error

Commit:
Replace `pg_ctl ... start` with directly calling `postgres`

Commit:
Code clean up + improve documentation

Commit:
Code cleanup + switch to cardano-serialization-lib

Commit:
Code cleanup + added tests for the isomorphism from csl Plutus data and
Pla Plutusdata

gnumonik commented on dens-query/api/postgres/dens.sql:
```diff
@@ -0,0 +1,116 @@
+-- References
```
Could you maybe write (or generate) a diagram for the DB schema? Mainly just for reference later on if we need it

jaredponn replied:
I see.

I'm going to hold on this until we get the changes discussed on slack in because that will shuffle the schema around a bit.

See #32

Commit:
Fixed buggy rollbacking which did not "undo" the actions that have
happened since the rollback

Commit:
Draft of database together

- Feature: inserting DensSetRows

- Feature: computing the greatest strict lower bound

- Tests: Integration tests with Postgres of the above two features

Commit:
Draft of dens-query

Commit:
Update `flake.nix`

Commit:
Update flake.lock + dens-query package.json

Commit:
Starting to add HTTP interface to the DB

Commit:
Reorganized `dens.sql`

Commit:
Added bare bones of an HTTP server interface to dens-query

Commit:
Added draft of the user manual for dens-query

Commit:
Changed build system of the user manual to reuse existing NPM
infrastructure, and refactored from `docs/*` --> `doc/*`

Commit:
Added support for both Unix domain sockets, and Internet domain sockets

Commit:
Shuffled around how the undo_log is freezed / unfreezed

Commit:
Removed @cardano-ogmios/client in favor of rolling ogmios interactions
ourselves, added the protocolNft to the database, and put infrastructure
s.t. it may be changed dynamically at runtime

Commit:
Added a "Set Protocol NFT" endpoint, and cleaned up docs

Commit:
Fixed documentation errors / added TODOs

Commit:
Fix assorted bugs + added a trivial `/api/check-health` (missing tests)

Commit:
Changed rollbacks to use the `block_slot` to roll back

Commit:
Modified chain synchronization s.t. it doesn't pipeline as much

Commit:
Improvied logging message

Commit:
Fixed bugs of foreign key violation when ogmios inserts the same block
twice, and turned the runtime depenendency of the dens.sql file in to a
derivation

Commit:
Fixed incorrect rollbacks (casting the bool `TRUE` to a `TEXT` is `'true'` not `'TRUE'`), added undo_logging of `UPDATE`, and fixed documentation typos

Commit:
Add custom error handler for dens-query s.t. it doesn't render a fancy
HTML page

Commit:
Fixed documentation error

Commit:
Added structured RRs to the database

Commit:
Added some verifications to ensure that only RRs with sane domain names
are added to the database

Commit:
Fixed long standing TODO of incorrectly computing the token name of the
ElementIdMintingPolicy

Commit:
Removed autogenerated file

Commit:
Fixed documentation error, and used the provided class name instead of
the hardcoded value when generating the ElementId token name

Commit:
Fixed typo

Commit:
Refactored internal names

Commit:
Fixed a variety of bugs in dens-query
- Improper setting of the protocol NFT
- Forgetting to add TxOutRefs to the database

Commit:
Attempt to fix illegal character in store path

gnumonik reviewed (approved):
Looks good!
